src/cth_readable_failonly.erl

```diff
@@ -236,50 +240,69 @@ handle_event(_, S) ->
 handle_info(_, State) ->
     {ok, State}.
 
-handle_call({lager, _} = Event, S=#eh_state{buf=Buf}) ->
+handle_call({lager, _} = Event, State) ->
     %% lager events come in from our fake handler, pre-filtered.
-    {ok, ok, S#eh_state{buf=[Event | Buf]}};
+    {ok, ok, buffer_event(Event, State)};
 handle_call({ct_pal, ignore}, S) ->
     {ok, ok, S};
-handle_call({ct_pal, _}=Event, S=#eh_state{buf=Buf}) ->
-    {ok, ok, S#eh_state{buf=[Event | Buf]}};
+handle_call({ct_pal, _}=Event, State) ->
+    {ok, ok, buffer_event(Event, State)};
 handle_call(ignore, State) ->
-    {ok, ok, State#eh_state{buf=[]}};
+    {ok, ok, State#eh_state{buf=queue:new(), stored_events=0}};
 handle_call(flush, S=#eh_state{buf=Buf}) ->
     Cfg = maybe_steal_logger_config(),
     ShowSASL = sasl_running() orelse sasl_ran(Buf) andalso S#eh_state.sasl,
     SASLType = get_sasl_error_logger_type(),
-    Buf =/= [] andalso io:put_chars(user, "\n"),
-    _ = [case T of
-            error_logger ->
-                error_logger_tty_h:write_event(Event, io);
-            sasl when ShowSASL ->
-                sasl_report:write_report(standard_io, SASLType, Event);
-            ct_pal ->
-                io:format(user, Event, []);
-            lager ->
-                io:put_chars(user, Event);
-            logger ->
-                Bin = log_to_binary(Event,Cfg),
-                io:put_chars(user, Bin);
-            _ ->
-                ignore
-         end || {T, Event} <- lists:reverse(Buf)],
-    {ok, ok, S#eh_state{buf=[]}}.
+    not queue:is_empty(Buf) andalso io:put_chars(user, "\n"),
+    flush(Buf, Cfg, ShowSASL, SASLType),
+    {ok, ok, S#eh_state{buf=queue:new(), stored_events=0}}.
 
 code_change(_, _, State) ->
     {ok, State}.
 
 terminate(_, _) ->
     ok.
 
+buffer_event(Event, S=#eh_state{buf=Buf, max_events=inf}) ->
+    %% unbound buffer
+    S#eh_state{buf=queue:in(Event, Buf)};
+buffer_event(Event, S=#eh_state{buf=Buf, max_events=MaxEvents, stored_events=StoredEvents}) when MaxEvents < StoredEvents ->
+    %% bound buffer; buffer not filled yet
+    S#eh_state{buf=queue:in(Event, Buf), stored_events=StoredEvents + 1};
+buffer_event(Event, S=#eh_state{buf=Buf0}) ->
+    %% bound buffer; buffer filled
+    {_, Buf1} = queue:out(Buf0),
+    S#eh_state{buf=queue:in(Event, Buf1)}.
+
+flush(Buf, Cfg, ShowSASL, SASLType) ->
+    case queue:out(Buf) of
+        {empty, _} -> ok;
+        {{T, Event}, NextBuf} ->
```

Comment:
The type signature here should make us expect `{{value,  {T, Event}}, NextBuf}`. You're going to get only ignored messages right now.